### DISCUSSION TITLE
How to set custom environment with Next 9.4?

### DISCUSSION THREAD

**Discussion by avxkim** [vercel/next.js#12737] [Help]
when i try:

```
NODE_ENV=staging yarn build && yarn start
```

but it gives me an error: "https://github.com/zeit/next.js/blob/master/errors/non-standard-node-env.md"

What's the correct way to set custom environments that differ from "development" and "production"?

**Comment by jamesmosier** [comment]
Since `NODE_ENV` is reserved by Next, you could instead [prefix them](https://nextjs.org/docs/basic-features/environment-variables) with `NEXT_PUBLIC_` if you needed different variable names/values.

If that's not helpful, what exactly are you looking for when you mentioned "custom environments"?

**Reply by avxkim** [reply]
I mean, currently we can use ".env.development" it corresponds to `yarn dev`, and ".env.production" that corresponds to `yarn build && yarn start`, what if i want the production build, but which uses ".env.staging" instead of ".env.production"?

**Comment by rafaelalmeidatk** [comment]
`NODE_ENV` is utilized to know if you are developing the app (so you can have dev features like HMR, for example) or building it for production (here, "production" means any environment that is not dev). If you want to set more environments you need to use a different env variable, I explained about it here: https://github.com/zeit/next.js/issues/10129#issuecomment-575921866

**Reply by avxkim** [reply]
Thanks, you mean i should create env called, like `PROJECT_ENV`, then start next this way:

```
PROJECT_ENV=staging yarn build && yarn start
```

But if i set custom variable, then how does it know which .env file to use? ".env.staging" or ".env.production"?

**Reply by rafaelalmeidatk** [reply]
Generally you do not have multiple .env files for each env ([ref](https://www.npmjs.com/package/dotenv#should-i-have-multiple-env-files)), instead you have only one .env file that has the values changed from environment to environment.

If you still want to have multiple .env files, the answer will depend on how you implemented their loading. Dotenv has a [`path`](https://www.npmjs.com/package/dotenv#path) field that allows you to load a specific .env file, you can use it to decide the path based on `PROJECT_ENV`. If you are using the new built-in Environment Variables (I haven't utilized it yet so someone please correct me if I am wrong), it seems that it doesn't support other envs besides `dev`, `prod` and `test` ([ref](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables)).

**Reply by avxkim** [reply]
I'm using a new built-in env variables, but they mention about `test` environment, but didn't said how? Since there's no `yarn test` command or something.

I just want to use envs, like in CRA apps, i usually have `.env.local`, `.env.staging`, `.env.development`, `.env.production` files. I choose which file to use this way:

```
"build:dev": "env-cmd -f .env.development yarn build",
"build:staging": "env-cmd -f .env.staging yarn build"
```

Can i do the same with Next?

Also i wonder, how does Next's 9.4 built-in env vars actually work. How do they pick which .env file to choose?

**Reply by rafaelalmeidatk** [reply]
> I'm using a new built-in env variables, but they mention about test environment, but didn't said how?

`test` is the third possible value of `NODE_ENV`, it is generally utilized by test runners like Jest/Mocha. You can test the load of the `.env.test` file with `NODE_ENV=test yarn build`

> Can i do the same with Next?

I am not sure you can use different environments other than the three I mentioned for the .env files with the current implementation.

> Also i wonder, how does Next's 9.4 built-in env vars actually work. How do they pick which .env file to choose?

It is explained in the doc page I linked: https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables

> The supported environments are development, production and test. The environment is selected in the following way:
> 
> `next dev` uses `development`
> `next build` and `next-start` use `production`

**Reply by avxkim** [reply]
> I am not sure you can use different environments other than the 3 I said for the .env files with the current implementation.

Why do they give a such example, like we can use any name as "[environment]"?

![image](https://user-images.githubusercontent.com/5739191/81646767-a3d10a80-944d-11ea-9d62-dbf2f394b073.png)

Also in a docs:

> Note: .env files should be included in your repository

If i include my `.env` file in a repo, then it would be used in any environment, that's why i thought if we can use different `.env.[envrionment]` files, then we must have a way to build a production app with custom environment names, it's confusing.

**Reply by rafaelalmeidatk** [reply]
> Why do they give a such example, like we can use any name as "[environment]"?

It means "`[environment]` here can be one of the three supported environments", they give an example slightly after: "For example: `.env.development`"

> Also in a docs:
> 
> Note: .env files should be included in your repository
> 
> If i include my .env file in a repo, then it would be used in any environment, that's why i thought if we can use different .env.[envrionment] files, then we must have a way to build a production app with custom environment names, it's confusing.

In this case, the `.env` file is for stuff like defaults configs and not secret data. The secrets are in `.env.local` that shouldn't be in your repo. From the docs:

> Note: .env files should be included in your repository, and .env*.local should be in .gitignore, as those files are intended to be ignored. Consider .local files as a good place for secrets, and non-local files as a good place for defaults.

**Comment by avxkim** [comment]
Well, with `env-cmd` i could actually set my custom environments, like this:

```
"build:staging": "env-cmd -f .env.staging yarn build && yarn start"
```

**Reply by ratkorle** [reply]
This works for me when using it locally. I have default .env file with environment for all stages and .env.staging on top of that.
But variables from .env.staging are not loaded when I deploy it on AWS Lambda@Edge. Do you have any suggestions?

**Comment by olaj** [comment]
I also have a bit of an issue with this.

Do i understand things correctly that i can't really use the .env files if i want to set different URLs and stuff for staging and so on? Isn't that a pretty common requirement? And if development and test is supported why couldn't staging be?

> instead you have only one .env file that has the values changed from environment to environment.

What does this comment mean? Can i change the values in the .env file somehow to accomplish this?

**Reply by harborhoffer** [reply]
It cascades. Put all your shared env vars in .env and override per environment with .env[environment]. Override per environment locally by appending .local

Still not sure I understand why it's encouraged to add .env files to the repo.

**Comment by jaybytez** [comment]
Same question and concern here.  We have development, integration, staging and production.  What is the correct way to supply the props?  I understand that the Next.js development/production concept is how the app is running and not necessarily where it is running.

**Comment by tunesmith** [comment]
Yes, I would also like a clearer answer on the nextjs best practice for how to specify staging-specific values.

**Comment by tamal-sen** [comment]
Looks like this thread is dead. I'd love to know more about how Nextjs team would address deployment with staging-specific values because I am trying to clean up my codebase and implement the built-in support for env

**Comment by zigang93** [comment]
it is quite frustrated when no best practices example to refer..
